* Results for 1: 645,466,4-6,454,931 (-)

** *Chromosome:* 1
** *Position:* 645,466,4-6,454,931 (268 nt)
** *Strand:* -
** *G+C:* 30.60 %
** *miRNA sequence:* UUAGAGAUUUUAGUUGGUUUAAUU
** *miRNA depth:* 44 (weigth: 44)
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
GUUAAACUUAAAUUGGCAUUGGUUUAGAGAUUUUAGUUGGUUUAAUUCAAUUUUACAAAA
((((.(((..((((((((.(((....(((((((..(((((((.((((((((((.((....

CUUGAUGGGUUAAUUGGGUAAACCAUUGAAACCAUUAACCAUUACAACCUAACUUAUUUU
............................................................

AUUCAUCAAACCAAUUGACUCAUCAAAUCAUUUGACUCAUCAACUCAUUUGAGUCAAAAU
............................................................

UUUCAACUCAUUACGGUUUAUGGGUUGAGUUGAAUUGAGUUGACCAAUGAAUUUUGACCC
............................)).))))))))))))))))))))))))...))

AUUUUGAUACCCCUAAUUGUAGUAUAGC
)...))......))))))..))).))))
=
** *Stability of the secondary structure of the precursor:* /MFE/ -23.95 kcal/mol | /AMFE/ -25.11 | /MFEI/ -0.82
** *Stability of the miRNA duplex (mirdup):* no
** *Total number of reads mapped to the precursor:* 204
** *Distribution of reads:* one island
** *miRBase alignment:* presence of alignments that cover the miRNA locus


* Results for 1: 651,467,7-6,514,930 (+)

** *Chromosome:* 1
** *Position:* 651,467,7-6,514,930 (254 nt)
** *Strand:* +
** *G+C:* 16.14 %
** *miRNA sequence:* CUUAUUUACAAUGUCAUGCUACU
** *miRNA depth:* 27 (weigth: 27)
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
GUAUUUUGUUAACAAUAUUUUAAAAUAAAGUUUUUUGUUUAAAUAAAUAUUGUUCUUACU
((((.....(((.(((((((.(((((((.(((((((((((....................

AAACGUUUUUUGUUUAAAUAAACAAAUCAUGUAUUUGAACUUAUUUACAAUGUCAUGCUA
............................................................

CUGACAUUAAUUAUUUAGGCAUGAAAAUAUUUAAUUAAAGAUUUAAAUCUUCUAAAUUCU
............................................................

UGCUUUUAGAGUUAUUAUAUCAUUAUUUAAAGAGGCAAAAAAUAUUAAUUUUUUAAAUAU
................................))))))))))).)))..)))).))))))

UAUUAUAUAUAUAU
).))).....))))
=
** *Stability of the secondary structure of the precursor:* /MFE/ -7.12 kcal/mol | /AMFE/ -14.80 | /MFEI/ -0.92
** *Stability of the miRNA duplex (mirdup):* no
** *Total number of reads mapped to the precursor:* 35
** *Distribution of reads:* one island
** *miRBase alignment:* presence of alignments, which do not overlap the miRNA locus


* Results for 1: 653,840,0-6,538,487 (-)

** *Chromosome:* 1
** *Position:* 653,840,0-6,538,487 (88 nt)
** *Strand:* -
** *G+C:* 38.64 %
** *miRNA sequence:* AGAUGAUGGGCUUAGAUGAUGGGC
** *miRNA depth:* 760 (weigth: 380)
** *Candidates with the same miRNA:* 1__6543374-6543768-6 
** *miRNA precursor:*
=
CCCUAGAAACAUCUAGAUGAUGGGCUUAGAUGAUGGGCCUUAUAAUGGGCAUCCAUCUUU
((((((.((((...((((((((((..((((.(((((((((......))))..))))))))

AUUCAUAACAUUUUUUUUGUUGUUGGGG
))))))..)))))....)))).))))))
=
This stem-loop structure is the MFE structure.
** *Stability of the secondary structure of the precursor:* /MFE/ -25.80 kcal/mol | /AMFE/ -29.32 | /MFEI/ -0.76
** *Stability of the miRNA duplex (mirdup):* yes
** *Total number of reads mapped to the precursor:* 1210
** *Distribution of reads:* two islands
** *miRBase alignment:* none


* Results for 1: 653,870,3-6,538,797 (+)

** *Chromosome:* 1
** *Position:* 653,870,3-6,538,797 (95 nt)
** *Strand:* +
** *G+C:* 27.37 %
** *miRNA sequence:* AUAGAUCGGUAAAAGCAUGUUUUC
** *miRNA depth:* 4 (weigth: 4)
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
GAUCGAUGAUUAUAGUAUAGAUCGGUAAAAGCAUGUUUUCAAUGAUUCAAUAAAGAAGAG
(((.(.(((((((((.((((((.((((..(((((((((((.....(((......))))))

AAUCAUGCUUAUCAUUUGUUUCUAUAAUUACUAUC
)).))))))))))))))))..)))))))))).)))
=
This stem-loop structure is the MFE structure.
** *Stability of the secondary structure of the precursor:* /MFE/ -22.00 kcal/mol | /AMFE/ -23.16 | /MFEI/ -0.85
** *Stability of the miRNA duplex (mirdup):* yes
** *Total number of reads mapped to the precursor:* 11
** *Distribution of reads:* random
** *miRBase alignment:* none


* Results for 1: 654,324,2-6,543,337 (+)

** *Chromosome:* 1
** *Position:* 654,324,2-6,543,337 (96 nt)
** *Strand:* +
** *G+C:* 36.46 %
** *miRNA sequence:* 
** *miRNA depth:*  (weigth: )
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
GAGUAAUUGUAUGUUGCACUCUUUUUCCCUUAACCAUGGUUUUUCCCAUUGGGUUUUCCU
(((((.(((.(((((((.(((.((...(((((.(((.((.....(((...)))....)))

GGUAAGGUUUUUAAUGAGGCAGCAUUUCAAAUACUU
))))))).....)).))))))))))..))).)))))
=
This stem-loop structure is the MFE structure.
** *Stability of the secondary structure of the precursor:* /MFE/ -30.00 kcal/mol | /AMFE/ -31.25 | /MFEI/ -0.86
** *Stability of the miRNA duplex (mirdup):* no
** *Total number of reads mapped to the precursor:* 24
** *Distribution of reads:* random
** *miRBase alignment:* none


* Results for 1: 654,329,1-6,543,423 (+)

** *Chromosome:* 1
** *Position:* 654,329,1-6,543,423 (133 nt)
** *Strand:* +
** *G+C:* 39.10 %
** *miRNA sequence:* 
** *miRNA depth:*  (weigth: )
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
UGGGUUUUCCUGGUAAGGUUUUUAAUGAGGCAGCAUUUCAAAUACUUUACGAGUUAUGAU
(((((((...((((..(((((((((((.((((.((((((((((((((..(..(..(((((

ACUAUAAUGGUCAUCAGGGGGAGUGUUAUGAAUAAAGAUGGAUGCCCAUUAUAAGGCCCA
.(......)))))))..)..))))))).))).....))))..))))))))).)))))).)

UCAUCUAAGCCCA
)))...)))))))
=
This stem-loop structure is the MFE structure.
** *Stability of the secondary structure of the precursor:* /MFE/ -36.40 kcal/mol | /AMFE/ -27.37 | /MFEI/ -0.7
** *Stability of the miRNA duplex (mirdup):* no
** *Total number of reads mapped to the precursor:* 261
** *Distribution of reads:* random
** *miRBase alignment:* none


* Results for 1: 654,337,4-6,543,768 (-)

** *Chromosome:* 1
** *Position:* 654,337,4-6,543,768 (395 nt)
** *Strand:* -
** *G+C:* 37.97 %
** *miRNA sequence:* AGAUGAUGGGCUUAGAUGAUGGGC
** *miRNA depth:* 760 (weigth: 380)
** *Candidates with the same miRNA:* 1__6538400-6538487-1 
** *miRNA precursor:*
=
UGUUUGAACAAAUGUGGAUCCAUUAUUGGGCUCAUUGAUGACUAUUGGAUUAGGGCCCAU
((((((((.(((.(((((((((((((.((((((((((.......................

UUGGUUCGCCGUUUGGCACGCUACAGCCACAAAUGACACGUGUAGGAGAGGAAGACGGAG
............................................................

UUAACGGGGAUGGUCACGGGAUCAUCCAAUGAACGUCAGUUAUUUUAAUGACUCAUUUGU
............................................................

AUUUGGCCACGUCAAAAUAUUUAGCAAAAUUCAAAUAUUUUAACUUCUUUUUUGUUAUAA
............................................................

ACAGAGAAAGAGAUAGAAGAAAUGAGUGUGUUAUUCCAUUGAGAUAAAGGGUUCUUUAUA
............................................................

UAGGAGUACAAGGAAAAGGGUAAAACCCUAGAAACAUCUAGAUGAUGGGCUUAGAUGAUG
.......................................................)))))

GGCCUUAUAAUGGGCAUCCAUCUUUAUUCAUAACA
))))).)))))))..)))))).))).)))).))))
=
** *Stability of the secondary structure of the precursor:* /MFE/ -26.26 kcal/mol | /AMFE/ -27.29 | /MFEI/ -0.72
** *Stability of the miRNA duplex (mirdup):* no
** *Total number of reads mapped to the precursor:* 1376
** *Distribution of reads:* random
** *miRBase alignment:* none


* Results for 1: 659,298,4-6,593,131 (+)

** *Chromosome:* 1
** *Position:* 659,298,4-6,593,131 (148 nt)
** *Strand:* +
** *G+C:* 36.49 %
** *miRNA sequence:* AAAGAUAAAGCGGAUUAAACUGUU
** *miRNA depth:* 12 (weigth: 12)
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
GACGGACAAAUCUGUUUAUGUUUGUAGUGUUGAGGACUCAAAAAAAAAAAACUCAUUGCU
(((((...(((((((((.(((((.(((.((.((.(((...........((((.((((((.

GACCAACUACAAACCGCUUUUACAUACAAAUAGAGUUGCUCCGCAAUGGUUUGUUGUCUA
((.(((((.........................))))).)).))))))))))))).)).)

CCCUAAAGAUAAAGCGGAUUAAACUGUU
).))).))))))))))))))...)))))
=
This stem-loop structure is the MFE structure.
** *Stability of the secondary structure of the precursor:* /MFE/ -34.20 kcal/mol | /AMFE/ -23.11 | /MFEI/ -0.63
** *Stability of the miRNA duplex (mirdup):* yes
** *Total number of reads mapped to the precursor:* 22
** *Distribution of reads:* one island
** *miRBase alignment:* none


* Results for 4: 50,512,-50,681 (+)

** *Chromosome:* 4
** *Position:* 50,512,-50,681 (170 nt)
** *Strand:* +
** *G+C:* 51.18 %
** *miRNA sequence:* 
** *miRNA depth:*  (weigth: )
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
GGUGAGCUGGACAAGCUAGGAUGAACUAGCUAGCUAAGUUAGGUGAGCUAAACCAGCUGG
(((((((.((((..((((((((((.(((.((((((..(((((..................

AAGACUAGCUAGCUCAGCUAGGUGAGCUGACCGACAGCUCGACUGAGCUCGGUGAGCUGA
............................................................

UCGUGUACUGACCCAGCUGGGUGGUCAUUCUAAGUAAGAUCUUGCUUACU
.......)))))..)))))).)))))))))).)))..).))).)))))))
=
** *Stability of the secondary structure of the precursor:* /MFE/ -36.38 kcal/mol | /AMFE/ -45.29 | /MFEI/ -0.89
** *Stability of the miRNA duplex (mirdup):* no
** *Total number of reads mapped to the precursor:* 21
** *Distribution of reads:* random
** *miRBase alignment:* none


* Results for 4: 77,043,-77,331 (-)

** *Chromosome:* 4
** *Position:* 77,043,-77,331 (289 nt)
** *Strand:* -
** *G+C:* 46.37 %
** *miRNA sequence:* 
** *miRNA depth:*  (weigth: )
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
UGGGCCUCUUCACCUAUUGCCAAUUGGUUUUAGGUUGGAAACCCACACUCUAAUAUGGUA
(((((.(((..((.((..((((((((((..(((((.(((((..(.((.(((.(((((.((

UCAGAGCCUGAUCCGCACAUACUCAACCCGAUCCACAUCGGUCCGGCCCAAUACUUGGCC
(((......(((((.....((((((((.((((............................

CGCCGAUCCAUGCCCAAACAUAUCGAGAUCAAUGGCUAGAAGAACCAUCAUCUCGAGGGG
............................................................

GCGUAUUAGGGACAAUGUCCCACAUCGAAAGUUGAGUAGGAUCUUAAGUGAUAUAUAUAA
.......................))))...))))))))))))).....)))))))))).)

GAUAUGAGCCUUUCCAUCUAUUGCCAAUUGGUUUUAUGUUGGAAGCCCA
))..)).)..))))))))))..))))))))))..)).)).))).)))))
=
** *Stability of the secondary structure of the precursor:* /MFE/ -59.19 kcal/mol | /AMFE/ -35.54 | /MFEI/ -0.77
** *Stability of the miRNA duplex (mirdup):* no
** *Total number of reads mapped to the precursor:* 213
** *Distribution of reads:* random
** *miRBase alignment:* none


* Results for 4: 77,276,-77,585 (+)

** *Chromosome:* 4
** *Position:* 77,276,-77,585 (310 nt)
** *Strand:* +
** *G+C:* 45.16 %
** *miRNA sequence:* 
** *miRNA depth:*  (weigth: )
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
AUAUUAGAGUGUGGGUUUCCAACCUAAAACCAAUUGGCAAUAGGUGAAGAGGCCCAUAUC
((((.(((.(((((((((((.(((((...((((((((...((.((...((((((((((..

UUAUAUAUACCACUUAAGAUCUUAUUCAACUUCCGAUGUGGGAUAUUAUCCCUAAUACGC
............................................................

CCCCUCGAGAUGAUGGUUCUUCUAGCCAUUGAUCUCGAUAUGUUUGGGCAUGGAUCGGCG
............................................................

GGCCAAGUAUUGGGCCGGACCGAUGUGGAUCGGGUUGAGUAUGUGCGGAUCGGGGUCUGA
............................................................

UACCAUAUCAGACUGUGGGCUUCCAACGUAAAACUAAUUGGCAAUAGGUGGAGAGGCUCA
.............))))))))))..)).))...))))))))...))))).).))))))))

UAUCUUAUAU
))))).))))
=
** *Stability of the secondary structure of the precursor:* /MFE/ -43.29 kcal/mol | /AMFE/ -36.81 | /MFEI/ -0.82
** *Stability of the miRNA duplex (mirdup):* no
** *Total number of reads mapped to the precursor:* 103
** *Distribution of reads:* random
** *miRBase alignment:* presence of alignments, which do not overlap the miRNA locus


* Results for 4: 2,979,01-,298,116 (-)

** *Chromosome:* 4
** *Position:* 2,979,01-,298,116 (216 nt)
** *Strand:* -
** *G+C:* 31.02 %
** *miRNA sequence:* 
** *miRNA depth:*  (weigth: )
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
CAUGUUUUAAAUGGUAGUUAACUCUUAGUUCAUGUUUAAAAUGGCCACAUAUAAAGUCGA
((((..(((((((.(((((((((.((((..((((..........................

UGUUUAAAAUGACCACAAUUUAAAACUUGACUAAUAAAGUCGUUAACGGAUGAUAACAGA
............................................................

AGUUAGAAGUUCCGUUAAUGACUGAUAACGGCAUGUCUAUUUUGGCCUGGUUAACGAAUA
............................................................

AUCAUGUUUUAAAAGUUAGUUAACGUUUAGUUCAUG
..))))..)))).))))))))).)))))))..))))
=
** *Stability of the secondary structure of the precursor:* /MFE/ -13.88 kcal/mol | /AMFE/ -24.54 | /MFEI/ -0.79
** *Stability of the miRNA duplex (mirdup):* no
** *Total number of reads mapped to the precursor:* 301
** *Distribution of reads:* random
** *miRBase alignment:* none


* Results for 4: 2,981,93-,298,561 (-)

** *Chromosome:* 4
** *Position:* 2,981,93-,298,561 (369 nt)
** *Strand:* -
** *G+C:* 30.89 %
** *miRNA sequence:* AAGCAGAACCCAAGAACAUAGUUU
** *miRNA depth:* 176 (weigth: 176)
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
GUUAAUGACUGCUAAUCUAUCCAAAACAACGUCGUUUGAGAAUAAUCGAAAGAACAUAAA
((((((((((.(((...(((((((((..((....(((((((((((((.((((((.((...

GAUUUUUUCCAUAUGAAAUUUCCAUAUCAAAAAGUUUUAAGCAGAACCCAAGAACAUAGU
(((((((((...((((((((((......................................

UUGAUAUAGAGAUAACCGCAGUGCCAGUAAUUGGAAGCAAAUAAAGGUGUUUUCGUUUAU
............................................................

AUAUUUGUUUUAAAUUUUAAAAGGAAUCAUGGAAGCAAAUUGAAGACUCAUUAUAAGAUU
............................................................

GUUCGUCUUGUUUGAGCGGUUGUAGAAAUAGAUUUGGGGAACUGGAGUUUCAUUUAGAAA
...........................................))))))))))...))))

ACAAUCCUGAUUUUCUUUGGAUUAUUCUCAAAACAAAGUCAUUUUAGGGAUACUAAUAGA
).))))...)).)))))).))))))))))))).....))..))))..)))))....))))

GUCGUUAAC
)))))))))
=
** *Stability of the secondary structure of the precursor:* /MFE/ -50.55 kcal/mol | /AMFE/ -23.93 | /MFEI/ -0.77
** *Stability of the miRNA duplex (mirdup):* no
** *Total number of reads mapped to the precursor:* 300
** *Distribution of reads:* random
** *miRBase alignment:* none


* Results for 4: 3,132,66-,313,379 (-)

** *Chromosome:* 4
** *Position:* 3,132,66-,313,379 (114 nt)
** *Strand:* -
** *G+C:* 29.82 %
** *miRNA sequence:* CUUGGUAUAUUCAAAUUAGGAUU
** *miRNA depth:* 11 (weigth: 11)
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
CUAUCAAAAACUUAAAACCUUUCAAAACCUCUUGGUAUAUUCAAAUUAGGAUUCUUCAUU
(((((((((.(.(((.((.((((((((((.(((((.(((......(((.((........)

CAUGAAAUCAUUAUUCCAAGGGUUUUGGGAUGUCUUACGAAAUAAUUUUGGUAG
).)))......))).))))))))))))))).)).))).)......)))))))))
=
This stem-loop structure is the MFE structure.
** *Stability of the secondary structure of the precursor:* /MFE/ -23.80 kcal/mol | /AMFE/ -20.88 | /MFEI/ -0.7
** *Stability of the miRNA duplex (mirdup):* no
** *Total number of reads mapped to the precursor:* 36
** *Distribution of reads:* random
** *miRBase alignment:* none


* Results for 4: 4,343,04-,434,483 (+)

** *Chromosome:* 4
** *Position:* 4,343,04-,434,483 (180 nt)
** *Strand:* +
** *G+C:* 31.11 %
** *miRNA sequence:* AGUGACCAUUGGAGCAUCGUGCCU
** *miRNA depth:* 72 (weigth: 72)
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
UGACUCUAUUUUAGAGGCUGCUCCAAUGGUCACAUGUAAAAUAGAAUUUUCUUUAAAAUA
(((((((((((((.(((((((((((((((((((...(((((((((.(((((((.((((((

GAGGAGAUAUCAUCUCUAUAAAUAAAGAUCUCUACAUUUAGUCUCUAUUUUUUUCCUAUU
(......................................................)))))

UUUGAGGAAACUCUAUUUUAAAAGUGACCAUUGGAGCAUCGUGCCUAUAAAAUAGGGUCA
)).))))))).)))))))))...)))))))))))))))....)))).)))))))))))))

=
** *Stability of the secondary structure of the precursor:* /MFE/ -66.37 kcal/mol | /AMFE/ -40.94 | /MFEI/ -1.32
** *Stability of the miRNA duplex (mirdup):* yes
** *Total number of reads mapped to the precursor:* 152
** *Distribution of reads:* two islands
** *miRBase alignment:* presence of alignments, which do not overlap the miRNA locus


* Results for 4: 6,182,59-,618,526 (-)

** *Chromosome:* 4
** *Position:* 6,182,59-,618,526 (268 nt)
** *Strand:* -
** *G+C:* 15.67 %
** *miRNA sequence:* 
** *miRNA depth:*  (weigth: )
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
AAAAAAUUGCAAGAUUAUCAUAUAAGUCACUCAACCAUAUACAGUAGAAACUCUAUAAAU
(((((((((((.((((((.(((((...((..((((........(((((((..........

UAAUAAUGUUGGGACCAUGAAAUUUUAUUAAUUUAUAGUGAUAUUAAUUUAUCCUAUAAA
............................................................

UUAAUAUUUAUUAAUUUAUCGUAUAAGUUAAUAAAUAUUAAUUUAUAAAAAUAUUUAUGU
............................................................

AAAAUACAAUACUAUAUUCUUUUUAAAAUGUUUUUAAUUAAUUUUCUAUAUGCUUUGUUG
..........................................))))))).......))))

UUUGUAUAUAUAAUUAUGUUAGUUUUUU
..))))))))))))).)).)))))))))
=
** *Stability of the secondary structure of the precursor:* /MFE/ -12.02 kcal/mol | /AMFE/ -19.74 | /MFEI/ -1.26
** *Stability of the miRNA duplex (mirdup):* no
** *Total number of reads mapped to the precursor:* 119
** *Distribution of reads:* random
** *miRBase alignment:* none


* Results for 4: 7,432,81-,743,392 (-)

** *Chromosome:* 4
** *Position:* 7,432,81-,743,392 (112 nt)
** *Strand:* -
** *G+C:* 22.32 %
** *miRNA sequence:* AUUUAGGGAAUCUAGUUUUUACUC
** *miRNA depth:* 19 (weigth: 19)
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
GAUUUUAAAAUUUUGGUUUCCCUACAAUUUGAAAAACUAUUUUGUUGUAAUUUUCCAUUC
((...(((((..((((.(((((((.((((((((((((((.....(((.((((........

AUGAUUUUAGUUUUUUAGUUUUUCUAAAAUUUAGGGAAUCUAGUUUUUACUC
..)))).))).....)))))))))..)))))))))))).)))).))))).))
=
This stem-loop structure is the MFE structure.
** *Stability of the secondary structure of the precursor:* /MFE/ -19.50 kcal/mol | /AMFE/ -17.41 | /MFEI/ -0.78
** *Stability of the miRNA duplex (mirdup):* no
** *Total number of reads mapped to the precursor:* 25
** *Distribution of reads:* one island
** *miRBase alignment:* none


* Results for 4: 7,433,86-,743,470 (-)

** *Chromosome:* 4
** *Position:* 7,433,86-,743,470 (85 nt)
** *Strand:* -
** *G+C:* 22.35 %
** *miRNA sequence:* AAGGUCUUGAAUGUUUUGUAGAUU
** *miRNA depth:* 31 (weigth: 31)
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
UAAGGUCUUGAAUGUUUUGUAGAUUUUGUAAGAGUUUUUGGAUUUUGUUUUUCAAAAACU
(((((((((((.((((.((.((((.......(((((((((((........))))))))))

UAUUUUCAUAACAAUCAAGAUUUUA
))))).)).)))).)))))))))))
=
This stem-loop structure is the MFE structure.
** *Stability of the secondary structure of the precursor:* /MFE/ -24.00 kcal/mol | /AMFE/ -28.24 | /MFEI/ -1.26
** *Stability of the miRNA duplex (mirdup):* yes
** *Total number of reads mapped to the precursor:* 44
** *Distribution of reads:* one island
** *miRBase alignment:* none


* Results for 4: 9,831,83-,983,345 (-)

** *Chromosome:* 4
** *Position:* 9,831,83-,983,345 (163 nt)
** *Strand:* -
** *G+C:* 23.31 %
** *miRNA sequence:* 
** *miRNA depth:*  (weigth: )
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
UCAAGUGAGAAACUAUGAGAUACUCUCUACAAGCCAUCAUUACAUUUCUCUUCUUUUAAU
(((((.((((((((..((((...((((.(((((...........................

UGAUUCAAUUUUUACACUAAAAUAAAAAUUUAAUCAUAUAAUAAAAUUAUAAUAAAAUAC
............................................................

UAAUAUUUUUCUUGUUCGAGAAACUUUUUGGGUUUCUCCUUGA
..........)))))..))))....)))).)))))))))))))
=
** *Stability of the secondary structure of the precursor:* /MFE/ -20.14 kcal/mol | /AMFE/ -18.77 | /MFEI/ -0.81
** *Stability of the miRNA duplex (mirdup):* no
** *Total number of reads mapped to the precursor:* 247
** *Distribution of reads:* random
** *miRBase alignment:* presence of alignments, which do not overlap the miRNA locus


* Results for 4: 9,832,92-,983,369 (+)

** *Chromosome:* 4
** *Position:* 9,832,92-,983,369 (78 nt)
** *Strand:* +
** *G+C:* 35.90 %
** *miRNA sequence:* AUGUAAUGAUGGCUUGUAGAGAGU
** *miRNA depth:* 279 (weigth: 279)
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
AGAAGAGAAAUGUAAUGAUGGCUUGUAGAGAGUAUCUCAUAGUUUCUCACUUGAGAAACU
((((((((............(..((((((((((.(((((.(((.....)))))))).)))

UUCUACACUCUCUCUUCU
)))))))..)))))))))
=
This stem-loop structure is the MFE structure.
** *Stability of the secondary structure of the precursor:* /MFE/ -25.80 kcal/mol | /AMFE/ -33.08 | /MFEI/ -0.92
** *Stability of the miRNA duplex (mirdup):* no
** *Total number of reads mapped to the precursor:* 567
** *Distribution of reads:* one island
** *miRBase alignment:* none


* Results for 4: 9,855,65-,985,737 (+)

** *Chromosome:* 4
** *Position:* 9,855,65-,985,737 (173 nt)
** *Strand:* +
** *G+C:* 36.42 %
** *miRNA sequence:* GAGAGAGAGAGAGAGAGAGAGAG
** *miRNA depth:* 4 (weigth: 0.12)
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
UUCUGUUCUUCUCUCUGUUUGUUUGAAUUCAGUAGCAAUUGUGUAUUGCGUCUUAACGUC
((((.((.((((((((.(((.((.(((((((((...........................

UUAAGCUAAUAUACCCUUAAGACUGUGUGAUUUUUCUUAGUUAUGAGUCGUAUUAAGAAU
............................................................

CCAAUUGCUUCGUUUUCUGAAUUGGAUUCCGAGAGAGAGAGAGAGAGAGAGAG
....................))))))))).)).))).)))))))).)).))))
=
** *Stability of the secondary structure of the precursor:* /MFE/ -24.93 kcal/mol | /AMFE/ -27.63 | /MFEI/ -0.76
** *Stability of the miRNA duplex (mirdup):* yes
** *Total number of reads mapped to the precursor:* 10
** *Distribution of reads:* one island
** *miRBase alignment:* none


* Results for 4: 100,111,5-1,001,209 (+)

** *Chromosome:* 4
** *Position:* 100,111,5-1,001,209 (95 nt)
** *Strand:* +
** *G+C:* 15.79 %
** *miRNA sequence:* 
** *miRNA depth:*  (weigth: )
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
GUAAAGAUGAUUUAAAAAUAUAUGAUAUUAUUUUAUUUGAUUUUAAAUGUAUAGUAAACU
(((.((((((((((((((((((((...((((((((((..((.......))..))))))..

GUGAGUUGUAUAUGUUUUAUUGAUAUUAUCUAUAU
))))....)))))))))..)))).))))))).)))
=
This stem-loop structure is the MFE structure.
** *Stability of the secondary structure of the precursor:* /MFE/ -12.50 kcal/mol | /AMFE/ -13.16 | /MFEI/ -0.83
** *Stability of the miRNA duplex (mirdup):* no
** *Total number of reads mapped to the precursor:* 164
** *Distribution of reads:* random
** *miRBase alignment:* none


* Results for 4: 102,384,7-1,024,244 (-)

** *Chromosome:* 4
** *Position:* 102,384,7-1,024,244 (398 nt)
** *Strand:* -
** *G+C:* 34.17 %
** *miRNA sequence:* 
** *miRNA depth:*  (weigth: )
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
AACUUUCAAAAAAUGGCCAAAUUAACCGUGAACUUUUGAAAUGGUCGUUUUAUACCUCAA
((((((.(((((.(((((((((((((((((((.(((((((((((((((((((((((((((

AAAAAAGUUGAUUUCUAAUUUAACCUAUAAUUUAUUGUUGACUUCUAAUUUAACAUAUAA
.(((((((((((((((((((((((((((((..............................

UUUAUCGUUGACCCAGCCAAAUAGACUCACCGUUAACACUCCUUAACAGCGCUCCUAACA
............................................................

GCCGUUACAAAACAUGGCCUAACAGCCGUUAUCGUUGGUUUAAGGACAAAAACAGUUGUU
............................................................

AGUUACGCUGUUAGGAGAGCUGUUAAGAAGUGUUAAUGGUGAGUCGAUUUGGCCGUGUCA
............................................................

ACGAUAACUUAUAGGUUAAAUUAGAAGUCAACUUUUUGUUGAGGUAUAAAACGGCCAUUU
........))))))))))))))))))))))))))))).))))))))))))))))))))))

CAAAAAUUCAUGGUUAAUUUGGCCACUUUUUAAAAGUU
))))).))))))))))))))))))).))))).))))))
=
** *Stability of the secondary structure of the precursor:* /MFE/ -110.57 kcal/mol | /AMFE/ -54.27 | /MFEI/ -1.59
** *Stability of the miRNA duplex (mirdup):* no
** *Total number of reads mapped to the precursor:* 364
** *Distribution of reads:* random
** *miRBase alignment:* presence of alignments, which do not overlap the miRNA locus


* Results for 4: 111,657,0-1,116,885 (-)

** *Chromosome:* 4
** *Position:* 111,657,0-1,116,885 (316 nt)
** *Strand:* -
** *G+C:* 28.48 %
** *miRNA sequence:* ACGAAUCAAACAGUGAAAAGUGAC
** *miRNA depth:* 604 (weigth: 302)
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
UUUAAGGGUUAAGUUGUAUACUAGUCACUUUUUAAAAUUUAAUUGAAAAAAUAGUCACUU
((((((((.....(((((...((((((((((((.(((((((((((..(((.(((((((((

UACUAUUUAUAAUUUGCAAAAUAGUCACUUUUCACUGUUUGAUUCGUACACUUAGUUUGU
((((((((((((((((((((.((((((((((((((((((((((((((((((((.(.((((

ACACCACUUUUAGUGAUCAGAUGUAUAUAAAAAAUACACAGCGUACACUUUACAUUUGUA
(((.((((((((.((..((((((((.........................))))))))..

CAAUAAAAGUGAUGUACAAUCCAAGUGUACGAAUCAAACAGUGAAAAGUGACUACUUUGC
)).)))))))).))))))).).)))))))))))))))))))))))))))))))).)))))

AAAUUAUGAAUAGUAAAGUGACUACUUUCCUAAUUAAGUUUGAAAAAGUGACUAUUUUGC
)))))))))))))))))))))))).)))..))))))))))).))))))))))))...)).

CAAAACACCCCUUGAA
))).....))))))))
=
This stem-loop structure is the MFE structure.
** *Stability of the secondary structure of the precursor:* /MFE/ -152.00 kcal/mol | /AMFE/ -48.10 | /MFEI/ -1.69
** *Stability of the miRNA duplex (mirdup):* yes
** *Total number of reads mapped to the precursor:* 1438
** *Distribution of reads:* one island
** *miRBase alignment:* presence of alignments that cover the miRNA locus


* Results for 4: 119,006,1-1,190,301 (-)

** *Chromosome:* 4
** *Position:* 119,006,1-1,190,301 (241 nt)
** *Strand:* -
** *G+C:* 24.07 %
** *miRNA sequence:* 
** *miRNA depth:*  (weigth: )
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
UAUUCCCUCCGUCCCACUAAGAUGAAUUUUUUAAAAAUUUCACAUUGUUUAAGAAAUAUU
(((((((((.((((((..(((((((((((((((...........................

GUAUAAAGACAAUUUUAACCUCUUAAUAUACUUUAAAAAUAAUAAUAAAUAUUUAUCUAA
............................................................

UUAUUUAGUAAGGAUAAAACGGGAAAACAGUUUUAAAAUUUGCAUUGGAAAUAUAAAACU
............................................................

UCUAUCUUUGUGGGACAAAUAUUUUUCCUAGAAAAUUUAUCUUUAUGGGACAGAGGGAGU
............................)))))))))))))))..)))))).))))))))

A
)
=
** *Stability of the secondary structure of the precursor:* /MFE/ -35.00 kcal/mol | /AMFE/ -23.69 | /MFEI/ -0.98
** *Stability of the miRNA duplex (mirdup):* no
** *Total number of reads mapped to the precursor:* 101
** *Distribution of reads:* random
** *miRBase alignment:* presence of alignments, which do not overlap the miRNA locus


* Results for 4: 121,201,8-1,212,144 (+)

** *Chromosome:* 4
** *Position:* 121,201,8-1,212,144 (127 nt)
** *Strand:* +
** *G+C:* 33.07 %
** *miRNA sequence:* 
** *miRNA depth:*  (weigth: )
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
UAUUGAACAUUGAGUAGGAUCUUAAGUGGUAUAUAGAAUAUGAGCCUCUCCACUCAUUGC
(((((((..(((((((((((((.((((((((((((..((((((((.((((((((((....

CAAUUGAUUUUGAGUGGAGAUGCUUAUAUCUUAUAUAUAUCACUUAGAUCCUACUCAAGU
..........)))))))))).))))))))....)))))))))))))))))))))))))..

UUCAAUG
)))))))
=
This stem-loop structure is the MFE structure.
** *Stability of the secondary structure of the precursor:* /MFE/ -66.10 kcal/mol | /AMFE/ -52.05 | /MFEI/ -1.57
** *Stability of the miRNA duplex (mirdup):* no
** *Total number of reads mapped to the precursor:* 94
** *Distribution of reads:* random
** *miRBase alignment:* none


* Results for 4: 123,091,5-1,231,313 (-)

** *Chromosome:* 4
** *Position:* 123,091,5-1,231,313 (399 nt)
** *Strand:* -
** *G+C:* 33.58 %
** *miRNA sequence:* AUCACAAUCGCGGUACAGUAAACG
** *miRNA depth:* 7 (weigth: 7)
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
AUCCGCGUUUAUUAUACCGCUAUUGUGAUUUACCCGACUAAUCCCACUCACACAAAAAAG
(((((((((((((.((((((.((((((((((((((((((((((((((.(...........

CAAUCUAUGUGAUCAAUGAGAGUAGUUCUAGUUCUUUAUUGAUUUAGAUUGUCAUUGGAA
............................................................

CAAUCUUCAACCUUUUUUGUCUAUCUUGGAUGAUGUAUUUUCGGAAAAUGAUUCUCUUUA
............................................................

AGAUGAUAUUAUUUAGGUUGACUGAUUGAUUUGGAGUCUUAUCUAUCCUUAUGAUUAGGA
............................................................

GUUUUAUCUAGUUCGAUUUGUCAUUUUCCUUACCAAUUAACACAACACAUAACAUAUAUA
............................................................

GAUAGAUUCACAUUAUAAAACAAAGACAAAACGACGUCGUUUCUUAACGGGUGUGGGAUU
..................................................).))))))))

AGUUGGGUAAAUCACAAUCGCGGUACAGUAAACGCGGAU
)))))))))))))))))).)))))).)))))))))))))
=
** *Stability of the secondary structure of the precursor:* /MFE/ -70.02 kcal/mol | /AMFE/ -33.20 | /MFEI/ -0.99
** *Stability of the miRNA duplex (mirdup):* yes
** *Total number of reads mapped to the precursor:* 36
** *Distribution of reads:* one island
** *miRBase alignment:* none


* Results for 4: 123,126,4-1,231,345 (+)

** *Chromosome:* 4
** *Position:* 123,126,4-1,231,345 (82 nt)
** *Strand:* +
** *G+C:* 40.24 %
** *miRNA sequence:* AAACGCGGAUUUGUGGUUAAUCUC
** *miRNA depth:* 48 (weigth: 24)
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
UGAGUGGGAUUAGUCGGGUAAAUCACAAUAGCGGUAUAAUAAACGCGGAUUUGUGGUUAA
(((((.((((((((((((..((((((((..(((..........)))....))))))))..

UCUCGAUUUAAUGACCGGUUCA
.))))).)))))..)).)))))
=
This stem-loop structure is the MFE structure.
** *Stability of the secondary structure of the precursor:* /MFE/ -22.20 kcal/mol | /AMFE/ -27.07 | /MFEI/ -0.67
** *Stability of the miRNA duplex (mirdup):* no
** *Total number of reads mapped to the precursor:* 69
** *Distribution of reads:* one island
** *miRBase alignment:* none


* Results for 4: 127,078,3-1,270,953 (-)

** *Chromosome:* 4
** *Position:* 127,078,3-1,270,953 (171 nt)
** *Strand:* -
** *G+C:* 47.95 %
** *miRNA sequence:* CGGAGAAGCUGUAGGCGCUUUAUC
** *miRNA depth:* 13 (weigth: 13)
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
GCGGAGAAGCUGUAGGCGCUUUAUCGAGAGGUUUUGUGUACGGGCUUCUGCUAAGGUUUG
(((((((((..(.(((((((...((((((((((...........................

CUCGUGGCAACUAGGAUGGGACUCCAACUCUUUGGUAUCUUCGAGUUGUGGAAAUCACGC
............................................................

AAAUUUCCGAGUUUAGCUUCUCGGAAUCGGUGUUUCUUCUCUUAUUUCCGC
..............))))))))))....))))))))..))....)))))))
=
** *Stability of the secondary structure of the precursor:* /MFE/ -23.20 kcal/mol | /AMFE/ -29.53 | /MFEI/ -0.62
** *Stability of the miRNA duplex (mirdup):* yes
** *Total number of reads mapped to the precursor:* 66
** *Distribution of reads:* one island
** *miRBase alignment:* none


* Results for 4: 143,878,0-1,438,882 (+)

** *Chromosome:* 4
** *Position:* 143,878,0-1,438,882 (103 nt)
** *Strand:* +
** *G+C:* 35.92 %
** *miRNA sequence:* UGUCCGUUGGAGAUUCUAUUGCCU
** *miRNA depth:* 7 (weigth: 7)
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
CAUUUUCUGGGCAAAUCAAAGAAAUUGGAAUCUUUAAUAGACUAAACAUUGAUGUCAAGG
((((.((.((((((...........(((((((((((((.(((.((.((((.(((((...)

AUAUGAGUGAUUGUCCGUUGGAGAUUCUAUUGCCUUGGUGAUG
)))).)))).))))).))))))))))))))))))).)).))))
=
This stem-loop structure is the MFE structure.
** *Stability of the secondary structure of the precursor:* /MFE/ -32.90 kcal/mol | /AMFE/ -31.94 | /MFEI/ -0.89
** *Stability of the miRNA duplex (mirdup):* yes
** *Total number of reads mapped to the precursor:* 17
** *Distribution of reads:* random
** *miRBase alignment:* none


* Results for 4: 143,920,3-1,439,307 (-)

** *Chromosome:* 4
** *Position:* 143,920,3-1,439,307 (105 nt)
** *Strand:* -
** *G+C:* 28.57 %
** *miRNA sequence:* 
** *miRNA depth:*  (weigth: )
** *Candidates with the same miRNA:* none
** *miRNA precursor:*
=
CCAAUAGUGAGAAGAGAAAUGUCAUGAUGGGUUGUAAAGAGUAUCUCAGAAUUAACUUUC
(((((((.(((((((((((((..((((((((..((((((.((.....((((......)))

UACACUUUAUUUCCGUCAUUUUCAUUAUUUUUUUUUUUUAUUGUG
))).)))))).))))))))...)))).))))))))).)))))).)
=
This stem-loop structure is the MFE structure.
** *Stability of the secondary structure of the precursor:* /MFE/ -24.30 kcal/mol | /AMFE/ -23.14 | /MFEI/ -0.81
** *Stability of the miRNA duplex (mirdup):* no
** *Total number of reads mapped to the precursor:* 98
** *Distribution of reads:* random
** *miRBase alignment:* none


